php/class-wp-https-detection.php

```diff
@@ -0,0 +1,144 @@
+<?php
+/**
+ * WP_HTTPS_Detection class.
+ *
+ * @package PWA
+ */
+
+/**
+ * WP_HTTPS_Detection class.
+ */
+class WP_HTTPS_Detection {
+
+	/**
+	 * The cron hook to check HTTPS support.
+	 *
+	 * @var string
+	 */
+	const CRON_HOOK = 'check_https_support';
+
+	/**
+	 * Option name for whether HTTPS is supported.
+	 *
+	 * @var string
+	 */
+	const HTTPS_SUPPORT_OPTION_NAME = 'is_https_supported';
+
+	/**
+	 * Secret for the HTTPS detection request.
+	 *
+	 * @var string
+	 */
+	const REQUEST_SECRET = 'https_detection_secret';
+
+	/**
+	 * Query arg key for the HTTPS detection token.
+	 *
+	 * @var string
+	 */
+	const REQUEST_TOKEN_QUERY_ARG = 'https_detection_token';
+
+	/**
+	 * Inits the class.
+	 */
+	public function init() {
+		add_action( 'wp', array( $this, 'schedule_cron' ) );
+		add_action( self::CRON_HOOK, array( $this, 'update_option_https_support' ) );
+		add_filter( 'cron_request', array( $this, 'ensure_http_if_sslverify' ), PHP_INT_MAX );
+	}
+
+	/**
+	 * Schedules a cron event to check for HTTPS support.
+	 */
+	public function schedule_cron() {
+		if ( ! wp_next_scheduled( self::CRON_HOOK ) ) {
+			wp_schedule_event( time(), 'hourly', self::CRON_HOOK );
+		}
+	}
+
+	/**
+	 * Makes a request to find whether HTTPS is supported, and stores the result in an option.
+	 */
+	public function update_option_https_support() {
+		update_option( self::HTTPS_SUPPORT_OPTION_NAME, $this->is_https_supported() );
+	}
+
+	/**
+	 * Makes a request to the home URL to determine whether HTTPS is supported.
+	 *
+	 * Returns true if the URL has the scheme of HTTPS, and it has the correct token.
+	 * The token is only to verify that the request is for the same site.
+	 * Also, in wp_remote_get(), 'sslverify' is true by default.
+	 *
+	 * @return boolean Whether HTTPS is supported.
+	 */
+	public function is_https_supported() {
+		$request = wp_remote_request(
+			add_query_arg(
+				self::REQUEST_TOKEN_QUERY_ARG,
+				$this->get_token(),
+				home_url( '/', 'http' )
+			),
+			array(
+				'timeout' => 10,
+				'headers' => array(
+					'Cache-Control' => 'no-cache',
+				),
+			)
+		);
+
+		if ( is_wp_error( $request ) || ! method_exists( $request['http_response'], 'get_response_object' ) ) {
+			return false;
+		}
+
+		$response   = $request['http_response']->get_response_object();
+		$parsed_url = wp_parse_url( $response->url );
+		if ( ! $response->success || ! isset( $parsed_url['query'] ) ) {
+			return false;
+		}
+
+		$query_args = wp_parse_args( $parsed_url['query'] );
+		$is_https   = (
+			isset( $parsed_url['scheme'], $query_args[ self::REQUEST_TOKEN_QUERY_ARG ] )
```

Comment:
Instead of looking at the URL's query vars, I think we should be looking at the body of the response. That is, look at `wp_remote_retrieve_body( $response )` for the expected token. A query param in the requested URL can be the signal for whether the response should include the ✅ indicator that we then look for here.